[BUG] the provider is crashed if config.ip_access_list is added

### Terraform CLI and Terraform TiDB Cloud Provider Version
<!---
Run `terraform version` to show the version, and paste the result for Terraform and the TiDB Cloud Provider between the ``` marks below.

If you are not running the latest version of Terraform and the TiDB Cloud Provider, please try to reproduce the issue in a more recent version in case it has been fixed already. 
-->
```
$ terraform version
Terraform v1.5.4
on darwin_arm64
+ provider registry.terraform.io/tidbcloud/tidbcloud v0.2.0
```
### Describe the bug
<!-- A clear and concise description of what the bug is. -->
After creating a Terraform file without `config.ip_access_list` and applying it, I then added `config.ip_access_list` to the file and applied it again. As a result, I encountered the following error.

```
$ terraform apply --auto-approve                 
tidbcloud_cluster.example_cluster: Refreshing state... [id=xxxx]

Terraform used the selected providers to generate the following execution plan. Resource actions are indicated with the following symbols:
  ~ update in-place

Terraform will perform the following actions:

  # tidbcloud_cluster.example_cluster will be updated in-place
  ~ resource "tidbcloud_cluster" "example_cluster" {
      ~ config           = {
          + ip_access_list = [
              + {
                  + cidr        = "0.0.0.0/0"
                  + description = "allow access from anywhere"
                },
            ]
          - paused         = true -> null
            # (3 unchanged attributes hidden)
        }
        id               = "xxxx"
        name             = "xxxx"
      ~ status           = {
          ~ cluster_status     = "AVAILABLE" -> (known after apply)
            # (2 unchanged attributes hidden)
        }
        # (5 unchanged attributes hidden)
    }

Plan: 0 to add, 1 to change, 0 to destroy.
tidbcloud_cluster.example_cluster: Modifying... [id=xxxx]
╷
│ Error: Plugin did not respond
│ 
│ The plugin encountered an error, and failed to respond to the plugin6.(*GRPCProvider).ApplyResourceChange call. The plugin logs may contain more details.
╵

Stack trace from the terraform-provider-tidbcloud_v0.2.0 plugin:

panic: runtime error: index out of range [0] with length 0

goroutine 54 [running]:
github.com/tidbcloud/terraform-provider-tidbcloud/internal/provider.clusterResource.Update({0x0?}, {0x105628040, 0x1400058aa80}, {{{{0x10562b6e8, 0x14000500870}, {0x10550ce60, 0x140005c44b0}}, {0x10562d568, 0x140000b6550}}, {{{0x10562b6e8, ...}, ...}, ...}, ...}, ...)
        github.com/tidbcloud/terraform-provider-tidbcloud/internal/provider/cluster_resource.go:628 +0xfd0
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).UpdateResource(0x14000224b00, {0x105628040, 0x1400058aa80}, 0x140001d9370, 0x1400047b280)
        github.com/hashicorp/terraform-plugin-framework@v1.1.1/internal/fwserver/server_updateresource.go:118 +0x534
github.com/hashicorp/terraform-plugin-framework/internal/fwserver.(*Server).ApplyResourceChange(0x105628040?, {0x105628040, 0x1400058aa80}, 0x1400035a320, 0x1400047b4a0)
        github.com/hashicorp/terraform-plugin-framework@v1.1.1/internal/fwserver/server_applyresourcechange.go:99 +0x11c
github.com/hashicorp/terraform-plugin-framework/internal/proto6server.(*Server).ApplyResourceChange(0x14000224b00, {0x105628040?, 0x1400058a930?}, 0x1400035a2d0)
        github.com/hashicorp/terraform-plugin-framework@v1.1.1/internal/proto6server/server_applyresourcechange.go:52 +0x1a8
github.com/hashicorp/terraform-plugin-go/tfprotov6/tf6server.(*server).ApplyResourceChange(0x1400021e140, {0x105628040?, 0x1400058a150?}, 0x140002f41c0)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov6/tf6server/server.go:816 +0x3c0
github.com/hashicorp/terraform-plugin-go/tfprotov6/internal/tfplugin6._Provider_ApplyResourceChange_Handler({0x1055f07e0?, 0x1400021e140}, {0x105628040, 0x1400058a150}, 0x140002f4000, 0x0)
        github.com/hashicorp/terraform-plugin-go@v0.14.3/tfprotov6/internal/tfplugin6/tfplugin6_grpc.pb.go:385 +0x174
google.golang.org/grpc.(*Server).processUnaryRPC(0x140002f21e0, {0x10562bd28, 0x1400011b040}, 0x14000590000, 0x140004d4090, 0x105b993a0, 0x0)
        google.golang.org/grpc@v1.51.0/server.go:1340 +0xb90
google.golang.org/grpc.(*Server).handleStream(0x140002f21e0, {0x10562bd28, 0x1400011b040}, 0x14000590000, 0x0)
        google.golang.org/grpc@v1.51.0/server.go:1713 +0x840
google.golang.org/grpc.(*Server).serveStreams.func1.2()
        google.golang.org/grpc@v1.51.0/server.go:965 +0x88
created by google.golang.org/grpc.(*Server).serveStreams.func1
        google.golang.org/grpc@v1.51.0/server.go:963 +0x298

Error: The terraform-provider-tidbcloud_v0.2.0 plugin crashed!

This is always indicative of a bug within the plugin. It would be immensely
helpful if you could report the crash with the plugin's maintainers so that it
can be fixed. The output above should help diagnose the issue.
```

### Steps to Reproduce
<!--
Please list the full steps required to reproduce the issue, for example:
1. `terraform init`
2. `terraform apply`
-->
1. `terraform init`
2. create a Terraform file without `config.ip_access_list`
3. `terraform apply`
4.  add `config.ip_access_list` to the file
5. `terraform apply`

--- a/internal/provider/cluster_resource.go
+++ b/internal/provider/cluster_resource.go
@@ -3,6 +3,7 @@ package provider
 import (
 	"context"
 	"fmt"
+	"sort"
 	"strings"
 
 	clusterApi "github.com/c4pt0r/go-tidbcloud-sdk-v1/client/cluster"
@@ -626,14 +627,54 @@ func (r clusterResource) Update(ctx context.Context, req resource.UpdateRequest,
 		return
 	}
 	if data.Config.IPAccessList != nil {
-		for index, key := range data.Config.IPAccessList {
-			if state.Config.IPAccessList[index].CIDR != key.CIDR || state.Config.IPAccessList[index].Description != key.Description {
+		// You cannot add an IP access list to an existing cluster without an IP rule.
+		if len(state.Config.IPAccessList) == 0 {
+			resp.Diagnostics.AddError(
+				"Update error",
+				"ip_access_list can not be added to the existing cluster.",
+			)
+			return
+		}
+
+		// You cannot insert or delete IP rule.
+		if len(data.Config.IPAccessList) != len(state.Config.IPAccessList) {
+			resp.Diagnostics.AddError(
+				"Update error",
+				"ip_access_list can not be changed, only components can be changed now",
+			)
+			return
+		}
+
+		// You cannot update the IP rule.
+		newIPAccessList := make([]ipAccess, len(data.Config.IPAccessList))
+		copy(newIPAccessList, data.Config.IPAccessList)
+		sort.Slice(newIPAccessList, func(i, j int) bool {
+			return newIPAccessList[i].CIDR < newIPAccessList[j].CIDR
+		})
+
+		currentIPAccessList := make([]ipAccess, len(state.Config.IPAccessList))
+		copy(currentIPAccessList, state.Config.IPAccessList)
+		sort.Slice(currentIPAccessList, func(i, j int) bool {
+			return currentIPAccessList[i].CIDR < currentIPAccessList[j].CIDR
+		})
+
+		for index, key := range newIPAccessList {
+			if currentIPAccessList[index].CIDR != key.CIDR || currentIPAccessList[index].Description != key.Description {
 				resp.Diagnostics.AddError(
 					"Update error",
 					"ip_access_list can not be changed, only components can be changed now",
 				)
 				return
 			}
+		}
+	} else {
+		// You cannot remove the IP access list.
+		if len(state.Config.IPAccessList) > 0 {
+			resp.Diagnostics.AddError(
+				"Update error",
+				"ip_access_list can not be changed, only components can be changed now",
+			)
+			return
 		}
 	}
 

